backend/optimizer.py

```diff
@@ -1,46 +1,127 @@
+        for s in self.producer.schedule:
+            self.differentiated_loads.append(self.differentiate_and_interpolate(s[1].load_profile, indices))
+
+        objective = np.zeros(len(self.producer.schedule))
+        # return optimize.minimize(self.to_minimize, objective, tol=1.0, method="trust-ncg")
+        configs = list(map(list, itertools.product([0, 1], repeat=len(self.differentiated_loads))))
+        return self.recursive_binary_optimizer(configs, float('inf'), configs[0])
+
+    def to_minimize(self, schedule: List[float]):
+        produced = []
+        consumed = []
+
+        for t in self.indices:
             consumed_t = 0
-            for i, j in enumerate(schedule):
-                if j > 0.0:
-                    consumed_t += load_profile_utils.power(self.producer.schedule[i][1].load_profile, t)
-            produced_t = load_profile_utils.power(self.producer.prediction, t)
-
-            produced += produced_t
-            consumed += consumed_t
-            if produced_t - consumed_t < 0:
-                penalty += 1
-
-        score = math.fabs(produced - consumed) + penalty * 0
-        return score
+            for i, load in enumerate(self.differentiated_loads):
+                if schedule[i] > 0:
+                    consumed_t += load[t]
+
+            consumed.append(consumed_t)
+            produced.append(self.differentiated_production[t])
+
+        penalty = 0
+        diff = 0
+        for p, c in zip(produced, consumed):
+            diff_t = abs(p - c)
+            diff += diff_t
+            if c > p:
+                penalty += diff_t
+
+        return diff, penalty
+
+    def binary_optimzier(self):
+        configs = map(list, itertools.product([0, 1], repeat=len(self.differentiated_loads)))
+        min_config = None
+        min_config_value = float('inf')
+        for c in configs:
+            diff, penalty = self.to_minimize(c)
+            if diff < min_config_value:
+                min_config_value = diff + penalty
+                min_config = c
+        return min_config
+
+    def recursive_binary_optimizer(self, configs, min_value, min_config):
+        if len(configs) == 0:
+            return min_config
+        current_config = configs[0]
+        diff, penalty = self.to_minimize(current_config)
+
+        if penalty > min_value:
+            config_indices = [i for i, x in enumerate(current_config) if x == 1.0]
+            new_configs = list(filter(lambda c: not all(map(lambda i: c[i], config_indices)), configs))
+            return self.recursive_binary_optimizer(new_configs, min_value, min_config)
+
+        elif penalty + diff < min_value:
+            return self.recursive_binary_optimizer(configs[1:], penalty + diff, current_config)
+
+        else:
+            return self.recursive_binary_optimizer(configs[1:], min_value, min_config)
+
+    def differentiate_and_interpolate(self, series: pd.Series, indices: List[int]):
```

Comment:
function to interpolate and differentiate a `pd.Series`, given `indices` to interpolate over.